### ISSUE TITLE
PostgREST configuration available in db_root_spec session

### ISSUE THREAD

**Issue by wayland** [PostgREST/postgrest#3029]
Hi!  

I'm helping with the effort to rewrite the OpenAPI stuff in SQL, in a separate module.  I wanted to bring to the attention of the PostgREST devs one of the features that we'll need to have added to PostgREST to complete this work.  

We need access to some of the configuration parameters from SQL.  Our thought is that PostgREST can run some `set_config` statements in SQL before calling pgrst.db_root_spec.  The parameters we currently need are:

Need:
- `pgrst.server_host` and `pgrst.server_port` (or alternately `server_proxy_uri` -- just let us know which)
- `pgrst.db_schemas`

Thanks!

**Comment by steve-chavez** [comment]
@wayland Would you like to help with this?

The configs are set here:

https://github.com/PostgREST/postgrest/blob/c3301a165377f218db0abed217cf29ea3bbd9254/src/PostgREST/Query.hs#L236-L258

**Comment by wayland** [comment]
Thanks!  Hopefully I'll get to this eventually.  

My goal is to hook jsonforms-editor ( https://jsonforms-editor.netlify.app/ ) to this, and hopefully, I'll be able to add queries to Postgres, and forms will appear in the webapp.  

The pieces I still need to hook together are:
- A build (docker-compose) that includes PostgREST, DataWharf, and some kind of ETL (maybe Apache Nifi?)
- Get OpenAPI 3.1+ working in Postgres
- A jsonforms-editor based app (to add to the build above)
- This ticket here (the last piece, because I've almost no experience with Haskell -- as a Perl programmer (with some Raku), though, I'm closer to functional languages than the average coder)

If I get the OpenAPI stuff finished, I might dive into this before the jsonforms stuff, though; hopefully it'll be a smaller task.

**Comment by plentydone** [comment]
Does anyone mind if I take a swing at this? This seems like a fairly small change. Am I missing any requirements beyond getting setPgLocals to set the new fields and seeing if I can scrounge together a test?

**Comment by laurenceisla** [comment]
> I'm personally not convinced that set_config is the right approach. I think we should do better for the root endpoint and use arguments to the respective function.

Agree. Not discarding this option altogether, but I'd like to try something like this first:

```sql
create or replace function root(
  server_host text,
  server_port int,
  db_schemas text[],
  version text
) as $$
  ...
$$;
```
```yaml
# postgrest.conf
db-root-spec = "root"
```

PostgREST would need to pass those values to the function.

**Comment by steve-chavez** [comment]
Hardcoding a function signature at this stage (db-root-spec is still experimental) seems more cumbersome than just passing the GUCs only for the root spec. Once `postgrest-openapi` stabilizes we can commit to a signature or perhaps just a json param.

**Comment by steve-chavez** [comment]
> create or replace function root(
  server_host text,
  server_port int,
  db_schemas text[],
  version text
) as $$

Even with a single json parameter, there was a lot of complexity when trying to special case discovery of the above proc. More details on https://github.com/PostgREST/postgrest/pull/5085#discussion_r3709074719.  I've settled on a single JSON GUC which seems the best interface as of now.

Ultimately I think we should try to obtain all the information from the `application_name`, with this we save on passing the same information on every request (which is wasteful both as function arguments or GUCs). On https://github.com/PostgREST/postgrest/issues/4863#issuecomment-4826633279 there's an idea to encode the host and port there and with https://github.com/PostgREST/postgrest/issues/5042 there could be a future where db-schemas is unnecessary.

But as it is now, we don't have those options and we should unblock this with the JSON GUC.
